sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,465 @@
+        this.dataSource = dataSource;
+        this.query = query;
+        this.username = username;
+        this.password = password;
+        this.preparedStatementSetter = preparedStatementSetter;
+        this.batchSize = batchSize;
+      }
+
+      public JdbcWriter<T> withDataSource(DataSource dataSource) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withQuery(String query) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withUsername(String username) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withPassword(String password) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withPreparedStatementSetter(
+          PreparedStatementSetter<T> preparedStatementSetter) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public JdbcWriter<T> withBatchSize(long batchSize) {
+        return new JdbcWriter<>(dataSource, query, username, password, preparedStatementSetter,
+            batchSize);
+      }
+
+      public void validate() {
+        Preconditions.checkNotNull(dataSource, "dataSource");
+        Preconditions.checkNotNull(query, "query");
+        Preconditions.checkNotNull(preparedStatementSetter, "preparedStatementSetter");
+      }
+
+      @Setup
+      public void connectToDatabase() throws Exception {
+        if (username != null) {
+          connection = dataSource.getConnection(username, password);
+        } else {
+          connection = dataSource.getConnection();
+        }
+        connection.setAutoCommit(true);
+      }
+
+      @StartBundle
+      public void startBundle(Context context) {
+        batch = new ArrayList<>();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        T record = context.element();
+
+        batch.add(record);
+        if (batch.size() >= batchSize) {
+          finishBundle(context);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle(Context context) throws Exception {
+        for (T record : batch) {
+          try {
+            PreparedStatement statement = connection.prepareStatement(query);
+            preparedStatementSetter.setParameters(record, statement);
+            try {
+              statement.executeUpdate();
+            } finally {
+              statement.close();
+            }
+          } catch (Exception e) {
+            LOGGER.error("Can't insert into database", e);
```

Comment:
Again: if data can't be inserted into the database, this is user data loss - it should be treated as a fatal error, not a log message. Just don't catch the exception.